aws/resource_aws_ec2_client_vpn_endpoint.go

```diff
@@ -0,0 +1,323 @@
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+			"connection_log_options": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"cloudwatch_log_group": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"cloudwatch_log_stream": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+					},
+				},
+			},
+			"dns_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsEc2ClientVpnEndpointCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	req := &ec2.CreateClientVpnEndpointInput{
+		ClientCidrBlock:      aws.String(d.Get("client_cidr_block").(string)),
+		ServerCertificateArn: aws.String(d.Get("server_certificate_arn").(string)),
+		TransportProtocol:    aws.String(d.Get("transport_protocol").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		req.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("dns_servers"); ok {
+		req.DnsServers = expandStringList(v.(*schema.Set).List())
+	}
+
+	if v, ok := d.GetOk("authentication_options"); ok {
+		authOptsSet := v.([]interface{})
+		attrs := authOptsSet[0].(map[string]interface{})
+
+		authOptsReq := &ec2.ClientVpnAuthenticationRequest{
+			Type: aws.String(attrs["type"].(string)),
+		}
+
+		if attrs["type"].(string) == "certificate-authentication" {
+			authOptsReq.MutualAuthentication = &ec2.CertificateAuthenticationRequest{
+				ClientRootCertificateChainArn: aws.String(attrs["root_certificate_chain_arn"].(string)),
+			}
+		}
+
+		if attrs["type"].(string) == "directory-service-authentication" {
+			authOptsReq.ActiveDirectory = &ec2.DirectoryServiceAuthenticationRequest{
+				DirectoryId: aws.String(attrs["active_directory_id"].(string)),
+			}
+		}
+
+		req.AuthenticationOptions = []*ec2.ClientVpnAuthenticationRequest{authOptsReq}
+	}
+
+	if v, ok := d.GetOk("connection_log_options"); ok {
+		connLogSet := v.([]interface{})
+		attrs := connLogSet[0].(map[string]interface{})
+
+		connLogReq := &ec2.ConnectionLogOptions{
+			Enabled: aws.Bool(attrs["enabled"].(bool)),
+		}
+
+		if attrs["enabled"].(bool) && attrs["cloudwatch_log_group"].(string) != "" {
+			connLogReq.CloudwatchLogGroup = aws.String(attrs["cloudwatch_log_group"].(string))
+		}
+
+		if attrs["enabled"].(bool) && attrs["cloudwatch_log_stream"].(string) != "" {
+			connLogReq.CloudwatchLogStream = aws.String(attrs["cloudwatch_log_stream"].(string))
+		}
+
+		req.ConnectionLogOptions = connLogReq
+	}
+
+	log.Printf("[DEBUG] Creating Client VPN endpoint: %#v", req)
+	resp, err := conn.CreateClientVpnEndpoint(req)
```

Comment:
It looks like we need to use a mutex and/or `resource.Retry()` to protect from some AWS concurrency errors:

```
--- FAIL: TestAccAwsEc2ClientVpnEndpoint_basic (4.44s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* aws_ec2_client_vpn_endpoint.test: 1 error occurred:
        	* aws_ec2_client_vpn_endpoint.test: Error creating Client VPN endpoint: OperationNotPermitted: Endpoint cannot be created while another endpoint is being created or the service linked role is being deleted
```

```suggestion
	var resp *ec2.CreateClientVpnEndpointOutput
	err := resource.Retry(1 * time.Minute, func() *resource.RetryError {
		var err error
		resp, err = conn.CreateClientVpnEndpoint(req)

		if isAWSErr(err, "OperationNotPermitted", "Endpoint cannot be created while another endpoint is being created") {
			return resource.RetryableError(err)
		}

		if err != nil {
			return resource.NonRetryableError(err)
		}

		return nil
	})
```

Reply:
Thanks for pointing this out!  This recommendation has been added.